Logging

robinandeer opened the pull request:
I should practise what I preach so here are some additions based on what we over at Clinical have picked up around logging best practices.

Basically we create a separate logger for each module for added granularity. Output looks like this:

```
2015-04-20 06:59:45,332 - taca.cli - INFO - starting CLI
```

I think it's more informative :smiley:

You also direct logging through the "root logger" created in the log module. This will control all logging throughout the package through inheritance and even loggers in dependencies if required - sweet!

Please test out how it now works by setting you log level to "DEBUG" in the config and running e.g. ``$ taca analysis`` - you should see a log event printed to STDERR!

guillermo-carrasco commented on taca/analysis/analysis.py:
```diff
@@ -1,19 +1,20 @@
 """ Analysis methods for TACA """
 import csv
+from datetime import datetime
```
that's a bit of a strange import organization (not saying that it wasn't before tohugh). I usually group total standard imports (`import whatever_standard_package`) at the beginning in alphabetical order, then total non-standard packages in alphabetical order as well, then relative standard imports and finally relative non-standard imports. 

Do you understand me? I am re-reading what I wrote and I'm not sure I would, haha

robinandeer replied:
Oh, but I think I understand :-)

Sorry I didn't plan to change so much apart from the logging stuff... My thinking is that it's simplest to group all standard lib imports together, then non-standard, and lastly intra-package imports.

If you think it's clearer to separate standard lib imports because some are "from this import that" than I will correct it!

guillermo-carrasco replied:
hmm is not very clear to me what [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) says about this :-/

It says standard imports first, but that's what I do as well, is just that I separate the total and partial imports..

guillermo-carrasco commented on taca/analysis/analysis.py:
```diff
@@ -359,7 +377,7 @@ def run_bcl2fastq(run, config):
         aggregated_tiles = cl_options.get('aggregated-tiles')
         if aggregated_tiles and aggregated_tiles in ['AUTO', 'YES', 'NO']:
-            cl.etend(['--aggregated-tiles', aggregated_tiles])
+            cl.extend(['--aggregated-tiles', aggregated_tiles])
```
oh well spotted!

robinandeer replied:
Don't thank me - PyLint would never let that slip through!

If your interested, I'm sure Atom can handle such a plugin *just as well* as Sublime, hehe :smirk:

guillermo-carrasco commented:
nice one, thank you for the code cleanup, I'm sure @vezzi will appreciate it :wink: hehe.

I'll give this a try just to make sure that everything works and that the log file keeps being written, other than that, looks awesome. 

Tack så jatte mycket! I'm so happy to have you guys collaborating ^^

robinandeer commented:
Awesome! Thank you :smiley: 

I have one more question. What are you using the file handler for? Wouldn't it be just simpler to let everything go to STDERR and redirect output on the command line to a file if needed? What's the benefit of an explicit file handler? I'm curious!

Galithil commented:
To me, the point of logging is that if you cut the power to the machine, you still have a trace of what happens. the only way to ensure that is to write to a file. 
STDERR is here to get error messages, not log messages. A lot of applications assume several things when they see a STDERR output. One example in case would be the Genologics Lims, if an EPP outputs something on STDERR (and the exit code != 0) the script will be marked as failed. 
Cron jobs also handle STDERR in a specific way, it is not a channel devoted to standard messages.

guillermo-carrasco commented:
exactly @Galithil . Also @robinandeer , another _practical_ benefit is that you ensure that you will always get the logs dumped to a file, avoiding the case of the person/cron that runs the software forgetting to redirect the output: was it `>`, or `&>`, or `1&2>`...? if you know what I mean ;-)

robinandeer commented:
Thanks guys!

@Galithil: I agree but I think another point of logging is to debug your code and then sending it to STDERR makes a lot of sense. But I guess just providing the option would be fine!

I'm not sure, however, whether a script should make assumptions of what/why stuff goes into STDERR. Doesn't seem very robust to me. I get the feeling STDERR is just as often treated as a channel for *diagnostics*. I'm not up-to-date on cronjobs and output streams. Care to fill me in?

@guillermo-carrasco: I guess I will probably find this out (or cause this) sooner or later, hehe

New question if you have time for a novice: how do I make a reasonable assumption of where to store the log-file by default? Current director? Name-spaced under $HOME? Should I bother learning about "syslog"?

robinandeer commented on taca/log/__init__.py:
```diff
@@ -2,14 +2,15 @@
-LOG.setLevel(logging.INFO)
+# get root logger
+ROOT_LOG = logging.getLogger()
+ROOT_LOG.setLevel(logging.INFO)
```
Doesn't this hardcode all INFO-messages and above to be printed to STDERR?

robinandeer commented on taca/log/__init__.py:
```diff
@@ -19,16 +20,16 @@
     log_level = LOG_LEVELS[log_level] if log_level in LOG_LEVELS.keys() else logging.INFO
 
-    LOG.setLevel(log_level)
+    ROOT_LOG.setLevel(log_level)
```
Actually here the log-level is updated from the hardcoded level set above, no? So everything printed to the file handler will also go to the stream handler? I'm confused...

guillermo-carrasco replied:
And isn't that a good behaviour :)? If you run TACA interactively you'll get the logs in the console (and in the file), and if you run TACA as a cronjob you'll still get the logs in the file.

robinandeer replied:
Yep! I'm glad where on the same page

guillermo-carrasco commented:
Nice @robinandeer !

Tried it out and works fine :) 

```
(master) ~/repos_and_code/TACA ((detached from clinical/logging)) ~> taca storage cleanup --site nas
2015-04-22 14:51:04,045 - taca.storage.storage - INFO - Moving old runs in /Users/guillem/tmp/

(master) ~/repos_and_code/TACA ((detached from clinical/logging)) ~> tail -n 1 ~/log/taca.log
2015-04-22 14:51:04,045 - taca.storage.storage - INFO - Moving old runs in /Users/guillem/tmp/
```

I'll wait for the last discussion about import order before merging :-D

robinandeer commented:
OK, I updated the imports to (what I believe) is your suggestion or the way it was before - I don't really mind.

I also rebased over the upstream master so it should be a (very) clean merge :smiley:

guillermo-carrasco commented:
ops @robinandeer , you broke the tests :-/

robinandeer commented:
Tests now pass locally.

Someone needs to look into the recent changes though. The most important one is that I didn't see any reason to pass around logger instances or assign them as class properties. The just live inside the scope of it's parent module.

This relates specifically to the ``TransferAgent`` class and it's "children". I can't see how my changes would affect that but still :smiley:

guillermo-carrasco commented:
It looks great. I appreciate your changes, looks much cleaner :-). Maybe @b97pla you want to have a look at Robin's last commit?

guillermo-carrasco commented:
this work fine, I'll merge :-)

b97pla commented:
@guillermo-carrasco sorry, just noticed I had a lot of old unread notifications but I see that you've already taken care of my old sins.. thanks!

guillermo-carrasco commented:
:-D
